src/filesystem/drive/inner.rs

```diff
@@ -48,11 +62,75 @@ impl InnerDrive {
             ))
     }
 
+    fn mark_ancestors_dirty(&mut self, node_id: NodeId) -> Result<(), OperationError> {
+        // Changes have happened in `node_id` walk up its parents to root marking nodes as dirty
+        let mut node_id = node_id;
+        self.dirty_nodes.push(node_id);
+        while let Some(parent_perm_id) = self.by_id(node_id)?.parent_id() {
+            // Setting a node's parent ID to its own ID is done in some operations such as moving a node
+            // to temporarily orphan it. Obviously this would cause an infinite loop and needs to be caught
+            if parent_perm_id == self.by_id(node_id)?.permanent_id() {
+                break;
+            }
+            node_id = self.lookup_internal_id(&parent_perm_id)?;
+            self.dirty_nodes.push(node_id);
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn clean_drive(&mut self) {
+        // Not being able to find a Node should not necessarily be an error
+        // Should this take PermanentId? Can the node Id shift out from underneath me?
+
+        // Take the dirty node list (replacing it with an empty Vec)
+        // Reverse the list and then remove duplicates
+        // (can't use Vec::dedup since that only removes consecutive duplicates)
+        let mut dirty = std::mem::take(&mut self.dirty_nodes);
+        dirty.reverse();
+        let mut node_list: Vec<_> = dirty.iter().unique().collect();
+
+        // Pop elements from back and update their size and Cid
+        while let Some(node_id) = node_list.pop() {
+            // Because of the work above we can assume that once we get here all of a nodes children are up to date
+            // The only error we can get from [`Self::by_id_mut`] is [`OperationError::InternalCorruption()`]
+            // which indicates the `NodeId` was not found. In our case that is fine, it just means it was deleted and
+            // we don't have to do anything
```

Comment:
This is an invariant I don't want to hold. We should return an InternalCorruption error in these cases. If a node is removed it's ID should also be removed from the dirty list. If we ever are expecting a NodeId to be valid and its missing that is a very bad error and it should not be hidden away silently

Reply:
If we modify `clean_drive()` to return a `Result` I think there would be no harm in using the question mark operator to return  any error we get  from `self.by_id` . 

The only case where this condition can happen I believe is if you did multiple calls to `rm` without cleaning in between (e.g. I `rm` a node, and then `rm` one of its ancestors, and then I clean).  In the future if we want to be able to batch up several operations before cleaning then we may have to solve this problem, by removing entries from the dirty list as they are removed from the filesystem

Reply:
Yeah I thought the whole purpose of tracking the dirty nodes outside of the node state, rather than immediately calculating and resolving the new values was to allow the aggregation of different operations before we "clean" the filesystem.

I suspect there is some easy ways to delay some of the calculation such as only calling the `clean_drive()` when we would read the `size()` or `cid()` on a dirty node. It would allow for things like mass renames in a directory without recalculating all its parents until we do a subsequent encode or display the contents of a directory (which would query for the size right now).

Reply:
The main reason is that as it is designed right now the only way we can know if a node has changed is by noting that we gave out a mutable reference to it, but that happens before the change occurs and there is no way to trigger an action when that reference gets dropped (You could maybe try to get fancy with the `Drop` trait? maybe?) .  A lot of the modifications one does with that mutable reference happen in setter functions where we could do the "cleaning" as the last step. However in the case of `Node::data_mut()` we are giving out a reference that is then directly operated over and we get the same issue. 

So for now it is the responsibility of functions that get one of those mutable references to call the clean function after they are done. 

All that to say I think this approach makes the most sense now, and I think with minor modifications we should be able to add optimizations like you called out to do mass-renamings and other compound operations with minimal calls to the clean function